File: app/src/main/java/com/ternaryop/photoshelf/importer/PostRetriever.java
package com.ternaryop.photoshelf.importer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import android.content.Context;
import android.support.annotation.NonNull;

import com.ternaryop.tumblr.Tumblr;
import com.ternaryop.tumblr.TumblrPost;
import io.reactivex.Observable;

public class PostRetriever {
    private final Tumblr sharedTumblr;
    private int offset;
    private int total;

    public PostRetriever(Context context) {
        sharedTumblr = Tumblr.getSharedTumblr(context);
    }

    public Observable<List<TumblrPost>> readPhotoPosts(final String tumblrName, final long lastPublishTimestamp) {
        return readPhotoPosts(tumblrName, lastPublishTimestamp, null);
    }

    public Observable<List<TumblrPost>> readPhotoPosts(final String tumblrName, final long lastPublishTimestamp, final String tag) {
        HashMap<String, String> params = buildParams(tag);

        offset = 0;
        total = 0;
        return Observable.generate(emitter -> {
            params.put("offset", String.valueOf(offset));
            List<TumblrPost> postsList = sharedTumblr.getPublicPosts(tumblrName, params);
            boolean loadNext = postsList.size() > 0;
            offset += postsList.size();

            final ArrayList<TumblrPost> newerPosts = new ArrayList<>();
            for (TumblrPost tumblrPost : postsList) {
                if (lastPublishTimestamp < tumblrPost.getTimestamp()) {
                    newerPosts.add(tumblrPost);
                } else {
                    loadNext = false;
                    break;
                }
            }

            total += newerPosts.size();

            emitter.onNext(newerPosts);
            if (!loadNext) {
                emitter.onComplete();
            }
        });
    }

    @NonNull
    private HashMap<String, String> buildParams(String tag) {
        HashMap<String, String> params = new HashMap<>();
        params.put("type", "photo");
        if (tag != null && !tag.trim().isEmpty()) {
            params.put("tag", tag);
        }
        return params;
    }

    public int getTotal() {
        return total;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/importer/PostRetriever.kt
package com.ternaryop.photoshelf.importer

import android.content.Context
import com.ternaryop.photoshelf.util.network.ApiManager
import com.ternaryop.tumblr.TumblrPost
import com.ternaryop.tumblr.android.TumblrManager
import io.reactivex.Observable

class PostRetriever(context: Context) {
    private val sharedTumblr = TumblrManager.getInstance(context)
    private val postManager = ApiManager.postManager(context)
    private var offset: Int = 0
    var total: Int = 0
        private set

    fun readPhotoPosts(tumblrName: String,
        lastPublishTimestamp: Long, tag: String? = null): Observable<List<TumblrPost>> {
        val params = buildParams(tag)

        offset = 0
        total = 0
        return Observable.generate { emitter ->
            if (offset == 0) {
                // TODO: Use PostManager API to update data on server, this will replace readPhotoPosts when migration from sqlite will be completed
                postManager.getLastPublishedTimestamp(tumblrName, lastPublishTimestamp)
            }
            params["offset"] = offset.toString()
            val postsList = sharedTumblr.getPublicPosts(tumblrName, params)
            var loadNext = postsList.isNotEmpty()
            offset += postsList.size

            val newerPosts = mutableListOf<TumblrPost>()
            for (tumblrPost in postsList) {
                if (lastPublishTimestamp < tumblrPost.timestamp) {
                    newerPosts.add(tumblrPost)
                } else {
                    loadNext = false
                    break
                }
            }

            total += newerPosts.size

            emitter.onNext(newerPosts)
            if (!loadNext) {
                emitter.onComplete()
            }
        }
    }

    private fun buildParams(tag: String?): HashMap<String, String> {
        val params = HashMap<String, String>()
        params["type"] = "photo"
        if (tag != null && !tag.trim { it <= ' ' }.isEmpty()) {
            params["tag"] = tag
        }
        return params
    }
}